Hotfix/2.2.6

poltak opened the pull request:
This contains:
- update to `syncClientLogEntry` schemas to add indices (assoc commit in `storex-sync`: https://github.com/WorldBrain/storex-sync/commit/471d82e75f9e80cf1deca86598ea34ef7778090e )
- quick-n-dirty migration to update newly indexed fields to ensure they always have a value
- moving the `runtime.onInstall` event set-up to happen earlier in the BG script setup (runs our on-install + update logic, like quick-n-dirty migrations). This one was really interesting. It indeed did not run all the time! I seem to have found the, recently introduced, spot that is responsible for stopping it running. More info in 8aca7da 
- update to Dexie. the intention is to see if this stops a bunch of `createObjectStore` errors many of our users are stuck with whenever interacting with their DBs

TODOs:

- [ ] test manual upgrades from a current `master` builds with **sync data** (@blackforestboi @ShishKabab could you assist here if you have some time today? Instructions are: https://github.com/WorldBrain/Memex/blob/master/docs/Testing-Ext-Updates.md)

poltak commented:
> test manual upgrades from a current master builds with sync data

I just did a test of this myself with 100+ pages, 20 records in `clientSyncLogEntry`. Everything worked great apart from one error that kept being thrown:
`Tried to toggle sync logging before logging middleware was created`

This was to do with the order of BG script setup calls that I changed in 8aca7da to fix the on-install/update logic being set up too late: the middleware setup needs to come before the rest of the sync setup (included in the on-install/update logic setup). Just pushed a fix in dc27a3a and it's all working now.

Everything else seems fine, but @ShishKabab can you think anything else that may become a problem now with the reordering of BG setup I have done in this work?

poltak commented:
I also just noticed lots of tests are failing on the CI with     
`Could not map following collection version undefined to application version: 2020-07-16T00:00:00.000Z`

I cannot reproduce locally - everything's passing there.

It seems to be something to do with the new collection version, being set to today's date. I'll try setting it back a day and see if that has any effect

ShishKabab commented:
It seems that this file is not up to date:
https://github.com/WorldBrain/Memex/blob/hotfix/2.2.6/src/sync/background/storage.ts

The application has application versions which are linear incrementing dates, but external modules also have their versions. So that file maps the module versions to application versions. Indeed I see the file being correct on this branch:
https://github.com/WorldBrain/Memex/blob/feature/sync-log-index/src/sync/background/storage.ts

poltak commented:
TODO

- [ ] mobile
  - [ ] quick-n-dirty migration
  - [ ] change schema over
  - [ ] update storage operations interfacing with changed collection
  - [ ] figure out if easy to do test update
  - [ ] test what happens with existing data on update
- [ ] ext
  - [ ] update storage operations interfacing with changed collection

poltak commented:
I finally sorted out those tests I was struggling with! This should now be all ready, at least on the extension side. Still need to go through some tasks for Memex Go. 

@blackforestboi I added the function to be able to seed some dummy data in `clientSyncLogEntry`. Though after doing it I realized it needs to be available before the update to actually seed the data, so it's not much use in here... Anyway I rebased it so it's now the first commit on the branch, so here's some instructions to test:

1. checkout the first commit of this branch 
    - via CLI, you can just enter `git fetch --all && git checkout 6a75a53`
    - via GUI, right click on the commit with message `Set up basic clientSyncLogEntry data seed function` and choose to checkout
2. `yarn && yarn build:watch`
2. get some data
    - in the BG console run `dataSeeders.seedClientSyncLogEntries({ amount: 1000 }).then(() => console.log('done'))` - `amount` is set to 1000 here to generate 1000 records, but you can set it to whatever. It will print `done` when finished seeding the collection
    - maybe run an import too to get some other data in there
3. checkout the latest commit of the branch (just checkout the branch as you normally do and by default it should put you on the latest commit)
2. `yarn && yarn build:watch`
2. press the "refresh"-looking button for Memex in `chrome:extensions` page
1. report back in here

blackforestboi commented:
Tested the upgrade and the real sync.
Seems to work all right for the upgrade process and the sync process too. In both ways actually. 
Can sync from mobile phone to computer and from computer to mobile without issues.
